Embed ICU via CMake

ppenzin opened the pull request:
Add CMake external project for ICU builds. Keep the script for downloading ICU in case it is needed for some of Visual Studio builds.

Add ICU target as a dependency for components which use ICU.

Closes #6424

ppenzin commented:
I am not certain this would pass CI yet, I still have issues with non-static build.

rhuanjl commented:
The icu download fails for me with this - keeps getting "retrying" messages but never succeeds.

ppenzin commented:
Oops, wrong issue in the commit message.

ppenzin commented:
ping @zenparsing

ppenzin commented:
Removed two vim history files checked in by accident

rhuanjl commented:
I've updated to cmake 3.17.1 and this now works - it was not working with cmake 3.7.1.

However still three possible issues:

1. CommonDefines.h attempts to include unicode/uvernum.h and it's possible to hit this point before icu is ready leading to an error and hence having to run the build command twice for success.

2. With the old way of downloading icu it was stored in deps/Chakra.ICU and then used for any build you did with the --embed-icu flag. With this method storing it in a subdirectory of the output file, if you do a test-build, a debug build and a release build you'll download and build ICU 3 times over.

3. The old method advised the user that it was downloading ICU and that this had a separate license to the rest of CC and asked them to confirm acceptance - this method doesn't have this notice.

ppenzin commented:
After a bit of thought I am pretty close to abandoning this. I will sleep on it, but I am not sure how to resolve the first point that you've made without major effort:

> 1. CommonDefines.h attempts to include unicode/uvernum.h and it's possible to hit this point before icu is ready leading to an error and hence having to run the build command twice for success.

Yes, I should have though about it. CommonDefines.h should not probably include this header, unless it uses something from it, but pulling it out would be tricky. There is way too many places where CommonDefines gets included (indirectly), and missing include in it would cause the build to fail even if that include is not warranted. The only solution for building ICU _before everything_ is to define wrappers for adding libraries and executables, which all would depend on ICU, but that is an overkill.

In the long run, we need to refactor how we include ICU (also see #6435), I'll try to investigate a bit more and document it.

> 2. With the old way of downloading icu it was stored in deps/Chakra.ICU and then used for any build you did with the --embed-icu flag. With this method storing it in a subdirectory of the output file, if you do a test-build, a debug build and a release build you'll download and build ICU 3 times over.
> 3. The old method advised the user that it was downloading ICU and that this had a separate license to the rest of CC and asked them to confirm acceptance - this method doesn't have this notice.

Both of those are easy to fix, we can pick the prefix for ICU to be in `deps` subdirectory, and add a message to the build that would alert the user that they are about to get ICU sources dropped in.

rhuanjl commented:
> After a bit of thought I am pretty close to abandoning this. I will sleep on it, but I am not sure how to resolve the first point that you've made without major effort:
> 
> > 1. CommonDefines.h attempts to include unicode/uvernum.h and it's possible to hit this point before icu is ready leading to an error and hence having to run the build command twice for success.
> 
> Yes, I should have though about it. CommonDefines.h should not probably include this header, unless it uses something from it, but pulling it out would be tricky. There is way too many places where CommonDefines gets included (indirectly), and missing include in it would cause the build to fail even if that include is not warranted. The only solution for building ICU _before everything_ is to define wrappers for adding libraries and executables, which all would depend on ICU, but that is an overkill.

I've just had a look and it seems that the only thing CommonDefines gets from that ICU header is the version number AND that is only then used in 2 places WScriptJsrt.cpp AND IntlEngineInterfaceExtensionObject.cpp, so one option would be to remove the include from CommonDefines and put it in those files instead I assume? Which could delay the point in the build where it's needed quite signifcantly.

Or could just have a message at the start when downloading ICU saying something like "Downloading ICU, the build may fail if the download is slow, if this happens please just retry"

rhuanjl commented:
Alternatively if this is to painful would it be possible to have a flag to embed ICU with the user providing a path to it?

For backwards compatability with build.sh it could do the download then provide this flag and the path; for a pure cmake build the user would need to build ICU (or locate it on their system first).

ppenzin commented:
Thank you, looks like I was able to get it unstuck with your help!

> I've just had a look and it seems that the only thing CommonDefines gets from that ICU header is the version number AND that is only then used in 2 places WScriptJsrt.cpp AND IntlEngineInterfaceExtensionObject.cpp, so one option would be to remove the include from CommonDefines and put it in those files instead I assume?

Turns out that there is a header for ICU manipulations, I've moved that lonely include there and added it to WScriptJsrt.cpp.

Also incorporated other comments - added a message about ICU having a different license and moved the build to `deps` subdirectory.

> Alternatively if this is to painful would it be possible to have a flag to embed ICU with the user providing a path to it?

There is a way to do that now, though I am not sure that is exposed as a `build.sh` option. This need to be cleaned up a bit though with the rest of ICU detection (future work).

ppenzin commented:
Great, include path is different between, MSVC and CMake. same file is able to find the include on Linux and OSX: https://chakrateam.visualstudio.com/ChakraCore%20CI/_build/results?buildId=47208

Need to think what I can do about this...

ppenzin commented:
It is working now, aside from CI getting stuck on Windows 8.

rhuanjl commented:
This still takes two tries to build from clean for me - Backend/AgenPeeps.cpp includes Backend/Backend.h which includes Runtime/PlatformAgnostic/ChakraPlatform.h which includes Runtime/PlatformAgnostic/UnicodeText.h which includes Runtime/PlatformAgnostic/ChakraICU.h and that tries to include unicode/uvernum.h and the build fails.

This happens about the same time the download finishes so I'm not sure if it's just a fraction of a second too soon.

ppenzin commented:
We should declare dependency on ICU for everything that needs it, but with the transitive includes like that it is still a bit tricky. If we don't we are bound to run into situations where it works on my machine but does not work on somebody else's. I'll see if I can understand what _really_ needs to depend on ICU, I also have a backup solution with declaring dependencies in `lib/CMakeLists.txt`, but it runs into a few issues of its own.

ppenzin commented:
@rhuanjl, this should do the trick with building components depending on ICU at the right time.

ppenzin commented:
@rhuanjl thank you!

@boingoing sorry to bug you again, but we have yet another instance of the same CI issue. Can you please restart it one more time? I'd be happy to kick the tires on it if I get access.

boingoing commented:
I kicked-off the failed jobs, @ppenzin. We're still having the conversation about how to provide the community access to the CI directly. Thank you for being patient with us on this one and I will try to be a little more vigilant in helping unblock this in PRs. As always, feel free to ping me about it. 👍

ppenzin commented:
@boingoing thank you very much!
